services/ui/src/components/Problems/index.js

```diff
@@ -1,229 +1,368 @@
+const Problems = ({problems}) => {
+  const { sortedItems, requestSort, getClassNamesFor } = useSortableProblemsData(problems);
+  const [severitySelected, setSeverity] = useState([]);
+  const [sourceSelected, setSource] = useState([]);
+  const [servicesSelected, setService] = useState([]);
 
-    const handleProblemFilterChange = (event) => {
-        setHasFilter(false);
+  const [problemTerm, setProblemTerm] = useState('');
+  const [hasFilter, setHasFilter] = useState(false);
+  const [problemStats, setProblemStats] = useState([]);
 
-        if (event.target.value !== null || event.target.value !== '') {
-            setHasFilter(true);
-        }
-        setProblemTerm(event.target.value);
-    };
+  const severities = getOptionsFromProblems(problems, 'severity');
+  const sources = getOptionsFromProblems(problems, 'source');
+  const services = getOptionsFromProblems(problems, 'service');
+
+
+  // Handlers
+  const handleProblemFilterChange = (event) => {
+    setHasFilter(false);
+
+    if (event.target.value !== null || event.target.value !== '') {
+      setHasFilter(true);
+    }
+    setProblemTerm(event.target.value);
+  };
+
+  const handleSourceChange = (source) => {
+    let values = source && source.map(s => s.value) || [];
+    setSource(values);
+  };
+
+  const handleSeverityChange = (severity) => {
+    let values = severity && severity.map(s => s.value) || [];
+    setSeverity(values);
+  };
+
+  const handleServiceChange = (service) => {
+    let values = service && service.map(s => s.value) || [];
+    setService(values);
+  };
+
+  // Options
+  const severityOptions = (severity) => {
+    return severity && severity.map(s => ({ value: s, label: s}));
+  };
+
+  const sourceOptions = (sources) => {
+    return sources && sources.map(s => ({ value: s, label: s}));
+  };
+
+  const serviceOptions = (services) => {
+    return services && services.map(s => ({ value: s, label: s}));
+  };
+
+  // Filters
+  const isFilteredBySeverity = (item) => {
+    return Object.keys(item).some(key => {
+      if (item[key] !== null) {
+        return severitySelected.indexOf(item['severity'].toString()) > -1;
+      };
+    });
+  }
+
+  const isFilteredbySource = (item) => {
+    return Object.keys(item).some(key => {
+      if (item[key] !== null) {
+        return sourceSelected.indexOf(item['source'].toString()) > -1;
+      };
+    });
+  }
+
+  const isFilteredByService = (item) => {
+    return Object.keys(item).some(key => {
+      if (item[key] !== null) {
+        return servicesSelected.indexOf(item['service'].toString()) > -1;
+      };
+    });
+  }
 
-    const handleSort = (key) => {
-        return requestSort(key);
+  const isFilteredByTextFilter = (item) => {
+    return Object.keys(item).some(key => {
+      if (item[key] !== null) {
+        return item[key].toString().toLowerCase().includes(problemTerm.toLowerCase());
+      }
+    });
+  }
+
+  const shouldItemBeShown = (item) => {
```

Comment:
Hi @timclifford 

As we discussed earlier - if you move the various checks you're doing here in the "if" statements' conditionals, we can simplify the logic of this function to something like (with the functions renamed)

const shouldItemBeShown = (item) => matchesSeveritySelector(item) && matchesServiceSelector(item) && matchesSourceSelector(item) && matchesTextFilter(item)

Reply:
Thanks again for looking at this - as discussed, renamed functions to something a little clearer and improved the conditional logic too!